### ISSUE TITLE
Post IDCT rounding needs to go - it induces noise & generational loss and takes extra time to compute

### ISSUE THREAD

**Issue by br3aker** [SixLabors/ImageSharp#1782]
New AAN IDCT implementation which I'm working on as part of jpeg decoder optimization/refactoring provides a faster and **more accurate** IDCT calculation. 

Current conversion pipeline has this "conform to libjpeg" step:
https://github.com/SixLabors/ImageSharp/blob/07d50d0e0cbe12be7425379195968584d7350106/src/ImageSharp/Formats/Jpeg/Components/Decoder/JpegBlockPostProcessor.cs#L76-L79

This does a huge quality hit (well, not really 'huge' if we are not talking about a couple of re-encodings) with noise. This is especially noticeable for so called 'generational loss'. Below there's a comparison of rounding vs no-rounding jpeg re-encoding without any change to the original image - no quality or subsampling changed.

**Original image**
![tree](https://user-images.githubusercontent.com/20967409/137999628-5f048671-5909-4a16-8c22-ec853c9a9b26.jpg)

**5 generations**
![gen5_comparison](https://user-images.githubusercontent.com/20967409/137999338-9671f00b-e4fd-449a-b348-e720bae89de2.png)

*Bottom row is a difference between original image and re-encoded one 

**25 generations**
![gen25_comparison](https://user-images.githubusercontent.com/20967409/137999763-cfa014a1-ebc2-403e-987c-6c553306ab0c.png)

---------------------------------------

Of course users won't re-encode their images 5 or even 25 times but this example shows the uselessness of IDCT rounding. Right now ImageSharp has a slower but more accurate IDCT which should be a 'feature', not a 'downside' which must be hidden. And on top of that it takes extra time to compute so it's a win-win :P

Issues like https://github.com/SixLabors/ImageSharp/issues/245 can't be solved simply because JFIF is a **lossy** image container - 1/255 pixel miss is an expected trade-off.

**Update**
Same applies to downscaling decoding from https://github.com/SixLabors/ImageSharp/pull/2076.

No-rounding code provides slightly better PSNR results but linked PR would provide IDCT algorithms with rounding before this issue is resolved.

@JimBobSquarePants @antonfirsov

**Comment by antonfirsov** [comment]
Related:

I just noticed https://github.com/SixLabors/ImageSharp/pull/1694#issuecomment-878989715 and 2e5b0ad74da6b476f812266b193be81c63ba17f5. As far as I remember, I used ImageMagick to generate the output, which uses libjpeg-turbo internally. Does this mean that the bug is actually not in our decoder but in libjpeg-turbo? Does photoshop maintain it's own codec?

**Comment by br3aker** [comment]
I have a strong feeling that those PNG's were saved with lossy compressions before entering optipng. You can the difference without any amplifying. Look for these images for example:

`.\ImageSharp\tests\Images\Input\Jpg\baseline\testorig.jpg`
`.\ImageSharp\tests\Images\External\ReferenceOutput\JpegDecoderTests\DecodeBaselineJpeg_testorig.png`

They have quite a difference, resaving initial jpeg to png via photoshop yield no difference.

>can you check for some example image if there is a difference between resaving with photoshop VS resaving with chrome? The latter definitely uses libjpeg-turbo with defaults.

But reference **png**(s) are spoiled, not jpeg(s). Chrome/PS would produce different results due to internal quality settings. But we want reference pixels for testing which is done via lossless png images which have a diff compared to initial jpegs.

Photoshop lossless png matters even on current master:
```
Current png reference
*** Jpg/baseline/testorig.jpg ***
Difference: 0,3754%
```
```
PS jpeg resave to lossless png
*** Jpg/baseline/testorig.jpg ***
Difference: 0,1936%
```

**Comment by JimBobSquarePants** [comment]
Waves Necromancer wand. 🪄

It's been 5 years since we looked at this but I'm on an issue bashing bust so....

Looking at the spec again, I do not think the extra rounding step is justified.

ISO/IEC 10918-1 explicitly defines rounding only as part of **quantization during encoding**:

```
Sq_uv = round(S_uv / Q_uv)
```

Where `S_uv` are the DCT coefficients and `Q_uv` is the quantization table. This produces the integer quantized coefficients that are written to the bitstream.

On the decoder side the process is simply the inverse:

```
R_uv = Sq_uv * Q_uv
```

This is dequantization. No rounding step is defined here because both `Sq_uv` and `Q_uv` are already integers.

After this the decoder performs the IDCT to reconstruct the spatial samples. The only unavoidable rounding occurs when converting the reconstructed samples to integer pixel values for the final image representation.

In other words, the spec requires rounding during **quantization**, not as a separate step in the decode pipeline after IDCT. The additional rounding historically present in some implementations appears to be an artifact of specific integer pipelines rather than something mandated by the standard.

Given that, I do not see a strong reason for ImageSharp to intentionally introduce additional rounding error purely to mimic legacy behavior. The decoder should simply perform dequantization and IDCT according to the spec and only round when converting the final sample values to the target pixel format.

Unless there is a concrete interoperability problem caused by removing the extra rounding, I would strongly lean towards simplifying the pipeline and removing it.

**Comment by antonfirsov** [comment]
Spec conformance is not the only aspect here. I think there should be an investigation on libjpeg-turbo's decoding settings related to this behavior, and understand how they are configured in popular software and why. I would assume that they default to an integer pipeline (making the rounding behavior our codec wants to conform to), we could check what happens by opting into the floating point one.

IMO the above is a natural step, because if we want to make this change, we can't escape the work of actually producing some reference images with a more standard conform configuration of a widely adapted external decoder (which is in fact libjpeg). Otherwise we would need to define the expected output of our decoder by the decoder itself - IMO this wouldn't be wise.

Also, user expectations may vary, we should research them and potentially keep an opt-in switch into the "legacy behavior".

**Comment by antonfirsov** [comment]
It might be even better to look into [IJG libjpeg](https://www.ijg.org/) (or [here](https://github.com/libjpeg-turbo/ijg/)), which is *the* reference implementation.

**Comment by JimBobSquarePants** [comment]
The reason libjpeg rounds is simply because it uses a fixed-point integer IDCT.

The original IJG implementation dates back to a time when floating point was either slow or unavailable on many systems, so the DCT/IDCT was implemented using scaled integers. In that model the math constantly rescales intermediate values with right shifts. A plain shift would truncate and introduce bias, so the code adds a bias before shifting, effectively rounding to nearest.

That rounding is therefore just a consequence of the integer arithmetic pipeline. It keeps the fixed-point math stable and unbiased when reducing precision between stages.

If the pipeline keeps higher precision (for example floats) until the final sample conversion, that intermediate rounding step simply is not needed.
